src/GUTI/GUTIAnnotatedTypeFactory.java

```diff
@@ -29,20 +43,65 @@ public GUTIAnnotatedTypeFactory(InferenceChecker inferenceChecker,
         postInit();
     }
 
+    @Override
+    public AnnotatedTypeMirror getAnnotatedType(Tree tree) {
+        AnnotatedTypeMirror atm = super.getAnnotatedType(tree);
+        if (atm instanceof AnnotatedPrimitiveType) {
+            GUTAnnotatedTypeFactory gutAnnotatedTypeFactory = (GUTAnnotatedTypeFactory) realTypeFactory;
+            Slot bottomConstantSlot = slotManager.getSlot(gutAnnotatedTypeFactory.BOTTOM);
+            AnnotationMirror bottomVarAnnot = slotManager.getAnnotation(bottomConstantSlot);
+            atm.replaceAnnotation(bottomVarAnnot);
+        }
+        return atm;
+    }
+
+    @Override
+    public TreeAnnotator createTreeAnnotator() {
+        return new ListTreeAnnotator(new ImplicitsTreeAnnotator(this), new GUTIInferenceTreeAnnotater(this,
+                realChecker, realTypeFactory, new GUTIVariableAnnotator(this, realTypeFactory, realChecker, slotManager, constraintManager), slotManager));
+    }
+/*    @Override
+    public AnnotatedTypeMirror getAnnotatedType(Tree tree) {
+        AnnotatedTypeMirror atm = super.getAnnotatedType(tree);
+        return atm;
+    }*/
+
     /**
      * The type of "this" is always "self".
      */
     @Override
     public AnnotatedDeclaredType getSelfType(Tree tree) {
         AnnotatedDeclaredType type = super.getSelfType(tree);
-        if (type != null) {
-            GUTAnnotatedTypeFactory gutATF = (GUTAnnotatedTypeFactory) InferenceMain
-                    .getInstance().getRealTypeFactory();
-            type.replaceAnnotation(gutATF.SELF);
+        GUTAnnotatedTypeFactory gutATF = (GUTAnnotatedTypeFactory) realTypeFactory;
```

Comment:
The try-block here is really ugly. Can you see whether this actually still crashes?
Add a big TODO that this should be cleaned up.
I'm not sure whether we need to create an equality constraint here. Maybe the old implementation was good enough?